pylint/extensions/private_import.py

```diff
@@ -0,0 +1,190 @@
+                "import-private-name",
+                node=node,
+                args=(imported_identifier, private_name_string),
+                confidence=HIGH,
+            )
+
+    @utils.check_messages("import-private-name")
+    def visit_importfrom(self, node: nodes.ImportFrom) -> None:
+        if utils.is_node_in_typing_guarded_import_block(node):
+            return
+        private_module_imports = self._get_private_imports([node.modname])
+        if private_module_imports:  # If module is private, do not check imported names
+            self.add_message(
+                "import-private-name",
+                node=node,
+                args=("module", private_module_imports[0]),
+                confidence=HIGH,
+            )
+        elif not self.same_root_dir(
+            node, node.modname
+        ):  # Only check imported names if the module is external
+
+            names = [n[0] for n in node.names]
+            private_names = self._get_private_imports(names)
+            if (
+                not private_names
+            ):  # No private names, don't have to check against typing annotations
+                return
+
+            # Only check names not used as type annotations
+            if not self.populated_annotations:
+                self._populate_type_annotations(
+                    node.root(), self.all_used_type_annotations
+                )
+                self.populated_annotations = True
+
+            private_names = [
+                n for n in private_names if n not in self.all_used_type_annotations
+            ]
+
+            if private_names:
+                imported_identifier = "objects" if len(private_names) > 1 else "object"
+                private_name_string = ", ".join(private_names)
+                self.add_message(
+                    "import-private-name",
+                    node=node,
+                    args=(imported_identifier, private_name_string),
+                    confidence=HIGH,
+                )
+
+    def _get_private_imports(
+        self,
+        names: List[str],
+    ) -> List[str]:
+        """Returns the private names from input names by a simple string check"""
+        private_names = [name for name in names if self._name_is_private(name)]
+        return private_names
+
+    @staticmethod
+    def _name_is_private(name: str) -> bool:
+        """Returns true if the name exists, starts with `_`, and if len(name) > 4
+        it is not a dunder, i.e. it does not begin and end with two underscores
+        """
+        return (
+            bool(name)
+            and name[0] == "_"
+            and (len(name) <= 4 or name[1] != "_" or name[-2:] != "__")
+        )
+
+    def _populate_type_annotations(
+        self,
+        node: nodes.LocalsDictNodeNG,
+        all_used_type_annotations: Set[str],
+    ) -> None:
+        """Adds into the set all_used_type_annotations the names of all names ever used as a type annotation
```

Comment:
Hmm I just saw that this only checks for "ever".

What about:
```python
import _private
my_var: _private.Thing = _private.Thing()
```

I know the typing doesn't make sense as it is implicit, but wouldn't this be allowed with the current checker?

Reply:
Right now this does emit, because in `import X; from X import Y` we only check typing usage on `Y`. This does seem like a frequent pattern and I'll check on `import X` too.

Besides that, a typing usage of a private import masks non-typing usages. It's a trade-off to miss those cases to not search everything for non-type usages.

We could efficiently check for your example since we already have a related node. On an `AnnAssign` annotated with the private type, check if there's a `target` and see that accesses the private module.

Reply:
What do you think about the typing/non-typing behavior @DanielNoord? I think besides this it's ready to merge.

Reply:
Sorry @areveny! This notification seems to have been lost in my mail, I never saw your reply.

I haven't checked this, but do both of these uses of `Thing()` show up in `for usage_node in node.locals[name]:`?
If so we could do something where `populate_` returns `bool` depending on whether it was used as typing and then do a last check in that `for` loop for whether that specific `usage_node` is typing related. If not > break/return.

I think we should be able to fix this trade-off before merging.